### PR TITLE
Jetty 9.4.x 3284 sessions and system clock jitter

### PR DESCRIPTION
**Pull request by janbartel**
#3284 

Fix the situation where the system clock (System.currentTimeMillis) is not consistent. This can lead to the session expiry timer going off *before* the calculated expiry time. The fix is essentially to restart the timer if the expiry time has not been reached when the timer fires. As part of this change, I also changed the way the expiry time is calculated:  before, it was only calculated when the session was accessed; now it is also recalculated when the last request drains out of the session.  This means previously, the time requests spent inside the session actually counted towards the expiry time, which is not quite correct; now, the time at which the session will expire is strictly from the time the session is idle.

### CODE REVIEW

File: jetty-server/src/main/java/org/eclipse/jetty/server/session/Session.java [hunk 1]
```diff
@@ -127,31 +126,45 @@ public void onTimeoutExpired()
...
 
-        public void schedule()
+        /**
+         * @param ms the timeout to set; -1 means that the timer will not be
+         *            scheduled
+         */
+        public void schedule (long time)
         {
-            if (_msec > 0)
+            if (time >= 0)
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/session/Session.java]: You don't want to schedule is `time==0`.

File: jetty-server/src/main/java/org/eclipse/jetty/server/session/Session.java [hunk 2]
```diff
@@ -559,31 +593,33 @@ public void updateInactivityTimer()
                 // sessions are not immortal
                 if (evictionPolicy == SessionCache.NEVER_EVICT)
                 {
-                    // timeout is just the maxInactive setting
-                    _sessionInactivityTimer.setTimeout(_sessionData.getMaxInactiveMs());
+                    // timeout is the time remaining until its expiry
+                    time = (remaining > 0? remaining:0);
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/session/Session.java]: Add spaces before and after operators `?` and `:`.

File: jetty-server/src/main/java/org/eclipse/jetty/server/session/Session.java [hunk 3]
```diff
@@ -559,31 +593,33 @@ public void updateInactivityTimer()
...
                     if (LOG.isDebugEnabled())
                         LOG.debug("Session {} evict on exit", getId());
                 }
                 else
                 {
                     // want to evict on idle: timer is lesser of the session's
-                    // maxInactive and eviction timeout
-                    _sessionInactivityTimer.setTimeout(Math.min(maxInactive, TimeUnit.SECONDS.toMillis(evictionPolicy)));
+                    // expiration remaining and the time to evict
+                    time = (remaining > 0? ( Math.min(maxInactive, TimeUnit.SECONDS.toMillis(evictionPolicy))) : 0);
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/session/Session.java]: Use consistent spacing, see above.

**Comment by janbartel** [comment]
@sbordet  I don't see how nanoTime can be used:  it is simply a magic number that can measure duration between two such magic numbers.  Duration is currently used only (with nanotime) for the timer that is measuring expiration time for sessions held in memory. We need to use millis in the persistent store (database/infinispan/mongo etc etc) because we need to:

* check on loading a session if it has already expired
* periodically execute queries to find sessions in the persistent store that have expired (they may no longer be in the active memory of any node, or alternatively may be in the active memory of more than one node and thus the persistent store has the most up-to-date expiry time).

So yes, while millis is subject to jitter, no problem with session expiration has ever been reported in all the years of the jetty project until 1 report last week. So this seems to be a very rare case.  This fix covers the case where the (nano based) timer expires before the (millis based) expiration time; timer expiry slightly after the expiration time is not a problem. Moreover, given the nature of nanoTime, and the nature of session expiry, I don't understand specifically what you are proposing: please elaborate.

**Comment by sbordet** [comment]
I guess what I am proposing is to use `System.currentTimeMillis()` only when you have to serialize `SessionData`. In all other cases use `System.nanoTime()`.

Create a session, store a transient nano time - no need to store the GMT time yet; checks on session expiration work with nano times; etc. 
Then you need to serialize the session. Only at that point you store a GMT time.
Deserialize the session in another JVM, store a transient nano time; same as above.

Basically you never compare against the GMT time unless to scavenge sessions, or check on deserialization. All the rest is using nano times.
Makes sense?

File: jetty-server/src/main/java/org/eclipse/jetty/server/session/Session.java (hunk 1 above)
Comment (janbartel) [reply jetty-server/src/main/java/org/eclipse/jetty/server/session/Session.java]: I disagree. This is protection in case the scheduler is ever called with 0 time remaining: in this case, we want the usual codepath followed to handle expirations, and that is for the timer to call onTimeoutExpired.

**Comment by janbartel** [comment]
I can't see nanoTime being a useful solution:

* the spec requires us to preserve creation time and last accessed time in millis, so we're already required to be millis based

* we have to store and use millis for scavenge across the persistence mechanism

* what we're discussing relates only to Session objects (that are not immortal and/or that are subject to cache eviction) in memory. As nanos is just a duration, to use it we'd have to add at least 2 nano fields to the Session object for expiry: one which is the System.nanoTime at which the session went idle (call it "idleNanos"), and one is the maxInactiveInterval (sec) converted to nanos (call it "idleDuration"). Then when the timer goes off, we'd have to compare if the current System.nanoTime - idleNanos >= idleDuration. Having done that, we mark the session as a candidate for scavenge, but note that the session will only be actually expired if we check in the persistence mechanism and find that the expiration time has passed, which is millis based anyway! So all this precision with the nano time is pretty pointless, because it all reduces to checking in the persistence mechanism against the precalculated millis expiry time!

So I think the most straightforward solution is to keep using millis to confirm expiry and simply re-adjust the existing timer if it goes off early.

Of course, feel free to make a more convincing argument for nanos.

**Comment by sbordet** [comment]
All right then.

**Comment by sbordet** [comment]
Please fix the code formatting, and fix the reason the build failed (javadocs, seems), then you can merge.
